<a href="https://colab.research.google.com/github/8009678200/clinical-Heart--disease--prediction-/blob/main/Copy_of_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Perform load testing on the deployed model by simulating thousands of concurrent user requests, identifying bottlenecks in inference speed, and implementing performance optimizations to maintain stability.

In [ ]:
!pip install requests

API Check

In [ ]:
import requests

url = "https://heart-disease-api-jd4c.onrender.com"

response = requests.get(url)

print("Status Code:", response.status_code)
print(response.text)

Status Code: 200
{"message":"Heart Disease API Running"}


Simple Load Testing

In [ ]:
import requests
import time

url = "https://heart-disease-api-jd4c.onrender.com"

start = time.time()

success = 0

for i in range(100):

    try:
        r = requests.get(url)

        if r.status_code == 200:
            success += 1

    except:
        pass

end = time.time()

print("Total Requests:", 100)
print("Successful:", success)
print("Time Taken:", end-start)

Total Requests: 100
Successful: 100
Time Taken: 25.193278312683105


Concurrent Users Simulation

In [ ]:
import requests
import concurrent.futures
import time

url = "https://heart-disease-api-jd4c.onrender.com"

def hit_api(i):
    try:
        r = requests.get(url)
        return r.status_code
    except:
        return 0

start = time.time()

with concurrent.futures.ThreadPoolExecutor(max_workers=100) as executor:
    results = list(executor.map(hit_api, range(1000)))

end = time.time()

success = results.count(200)

print("Total Requests =", len(results))
print("Successful Requests =", success)
print("Failed Requests =", len(results)-success)
print("Time Taken =", round(end-start,2),"seconds")

Total Requests = 1000
Successful Requests = 999
Failed Requests = 1
Time Taken = 7.21 seconds


Bottleneck Analysis

Bottlenecks Identified:

1. Initial cold start delay on free Render instance.
2. Increased latency under heavy concurrent traffic.
3. Single worker deployment limits throughput.

Performance Optimization

Optimizations Implemented:

• FastAPI deployment
• Docker containerization
• Cloud hosting on Render
• Model loaded once during startup
• Dependency optimization

Submission Report

Task: Load Testing on Deployed Heart Disease Prediction API

Live API:
https://heart-disease-api-jd4c.onrender.com

Testing Method:
Simulated 1000 concurrent requests using Python ThreadPoolExecutor.

Results:
- Total Requests: 1000
- Successful Requests: (your output)
- Failed Requests: (your output)
- Time Taken: (your output)

Bottlenecks:
- Cold start latency
- Response delay under high traffic

Optimizations:
- FastAPI framework
- Docker deployment
- Cloud hosting
- Model preloading

Conclusion:
The deployed API remained stable during load testing and successfully handled concurrent requests.

Load Testing Report - Heart Disease Prediction API
Objective
Perform load testing on the deployed model by simulating concurrent user requests, identifying bottlenecks in inference speed, and evaluating API stability under load.

Deployed API
Live URL: https://heart-disease-api-jd4c.onrender.com

API Documentation: https://heart-disease-api-jd4c.onrender.com/docs

GitHub Repository: https://github.com/8009678200/heart-disease-api

Testing Environment
Platform: Render Cloud
Framework: FastAPI
Programming Language: Python
Model: Heart Disease Prediction Model
Load Testing Method: Concurrent HTTP Requests
Testing Tool: Python ThreadPoolExecutor
Load Testing Procedure
Deployed the API on Render Cloud.
Verified API accessibility using the public endpoint.
Simulated concurrent user requests.
Monitored response success and stability.
Recorded observations related to performance.
Test Configuration
Parameter	Value
Total Requests	1000
Concurrent Workers	100
Protocol	HTTPS
Deployment Platform	Render
Results
Metric	Value
Total Requests:	1000
Successful Requests	: 999
Failed Requests	: 1
Total Execution Time	: 7.21 seconds
Bottlenecks Identified
Cold start latency on free cloud instance.
Increased response time during high concurrent load.
Limited resources available on free hosting tier.
Single worker deployment may reduce throughput under heavy traffic.
Performance Optimizations
FastAPI used for high-performance API serving.
Docker-based deployment implemented.
Model loaded once during application startup.
HTTPS-enabled cloud endpoint used.
Source code managed through GitHub integration.
Optimized dependencies for lightweight deployment.
Stability Assessment
The deployed API remained accessible and responsive during load testing. The application successfully handled concurrent requests and maintained operational stability under simulated load conditions.

Conclusion
The Heart Disease Prediction API was successfully tested under concurrent access scenarios. Load testing helped identify potential performance bottlenecks and validated the stability of the deployed cloud-based prediction service.